File: apps/desktop/src/features/chart/ultimateSupportResistance/usrTimeframe.ts
import type { UsrSettings } from './usrSettings';
import type { UsrAnalysisCandle, UsrCandle, UsrComputeContext } from './usrTypes';
import { applyLaggedVolumeBaselines, atrSeries, isFiniteCandle, isRegularSession } from './usrMath';

const MONDAY_EPOCH_OFFSET = 345_600;
const DAY = 86_400;
const WEEK = 604_800;

export interface UsrPreparedHistory {
  /** Canonical chart input, including the newest open candle for rendering. */
  presentationCandles: UsrCandle[];
  /** Fully closed chart candles used by every analytical state transition. */
  chartCandles: UsrCandle[];
  analysisCandles: UsrAnalysisCandle[];
  analysisSeconds: number | null;
  timeframeTag: string;
  usedChartTimeframe: boolean;
  warnings: string[];
}

interface TimeframeValue {
  seconds: number | null;
  months: number | null;
  ticks: number | null;
  tag: string;
}

function fixedSeconds(seconds: number): TimeframeValue {
  return { seconds, months: null, ticks: null, tag: String(seconds) };
}

function fixedMonths(months: number): TimeframeValue {
  return { seconds: null, months, ticks: null, tag: `${months}M` };
}

function fixedTicks(ticks: number): TimeframeValue {
  return { seconds: null, months: null, ticks, tag: `${ticks}T` };
}

function autoTimeframe(chartSeconds: number): TimeframeValue {
  if (chartSeconds <= 5) return { ...fixedSeconds(15), tag: '15S' };
  if (chartSeconds <= 15) return { ...fixedSeconds(60), tag: '1' };
  if (chartSeconds <= 30) return { ...fixedSeconds(120), tag: '2' };
  if (chartSeconds <= 60) return { ...fixedSeconds(300), tag: '5' };
  if (chartSeconds <= 180) return { ...fixedSeconds(900), tag: '15' };
  if (chartSeconds <= 300) return { ...fixedSeconds(1_800), tag: '30' };
  if (chartSeconds <= 900) return { ...fixedSeconds(3_600), tag: '60' };
  if (chartSeconds <= 1_800) return { ...fixedSeconds(7_200), tag: '120' };
  if (chartSeconds <= 3_600) return { ...fixedSeconds(14_400), tag: '240' };
  if (chartSeconds <= 7_200) return { ...fixedSeconds(28_800), tag: '480' };
  if (chartSeconds <= 14_400) return { ...fixedSeconds(DAY), tag: '1D' };
  if (chartSeconds <= 43_200) return { ...fixedSeconds(2 * DAY), tag: '2D' };
  if (chartSeconds <= DAY) return { ...fixedSeconds(WEEK), tag: '1W' };
  if (chartSeconds <= 3 * DAY) return { ...fixedSeconds(2 * WEEK), tag: '2W' };
  if (chartSeconds <= WEEK) return fixedMonths(1);
  if (chartSeconds <= 2 * WEEK) return fixedMonths(2);
  if (chartSeconds <= 31 * DAY) return fixedMonths(3);
  return fixedMonths(12);
}

export function parseUsrTimeframeValue(value: string): TimeframeValue | null {
  const match = /^([1-9]\d*)?(T|S|D|W|M)?$/.exec(value.trim().toUpperCase());
  if (!match) return null;
  const unit = match[2] ?? '';
  if (!match[1] && unit.length === 0) return null;
  const amount = Number(match[1] ?? 1);
  if (!Number.isSafeInteger(amount)) return null;
  switch (unit) {
    case 'T':
      return [1, 10, 100, 1_000].includes(amount) ? fixedTicks(amount) : null;
    case 'S':
      return [1, 5, 10, 15, 30, 45].includes(amount)
        ? { ...fixedSeconds(amount), tag: `${amount}S` }
        : null;
    case 'D':
      return amount <= 365 ? { ...fixedSeconds(amount * DAY), tag: `${amount}D` } : null;
    case 'W':
      return amount <= 52 ? { ...fixedSeconds(amount * WEEK), tag: `${amount}W` } : null;
    case 'M':
      return amount <= 12 ? fixedMonths(amount) : null;
    default:
      if (!match[1]) return null;
      return amount <= 1_440 ? { ...fixedSeconds(amount * 60), tag: String(amount) } : null;
  }
}

function selectedTimeframe(
  settings: UsrSettings,
  chartSeconds: number | null,
): TimeframeValue | null {
  switch (settings.analysisTimeframe) {
    case 'chart':
      return chartSeconds === null
        ? null
        : {
            ...fixedSeconds(chartSeconds),
            tag: chartTimeframeTag(chartSeconds) ?? String(chartSeconds),
          };
    case 'auto':
      return chartSeconds === null ? null : autoTimeframe(chartSeconds);
    case '4h':
      return { ...fixedSeconds(14_400), tag: '240' };
    case '1d':
      return { ...fixedSeconds(DAY), tag: '1D' };
    case '3d':
      return { ...fixedSeconds(3 * DAY), tag: '3D' };
    case '1w':
      return { ...fixedSeconds(WEEK), tag: '1W' };
    case '2w':
      return { ...fixedSeconds(2 * WEEK), tag: '2W' };
    case '1m':
      return fixedMonths(1);
    case 'custom':
      return parseUsrTimeframeValue(settings.customTimeframe);
  }
}

function monthIndex(epochSeconds: number): number {
  const date = new Date(epochSeconds * 1_000);
  return date.getUTCFullYear() * 12 + date.getUTCMonth();
}

function chartTimeframeTag(seconds: number | null): string | null {
  switch (seconds) {
    case 60:
      return '1';
    case 300:
      return '5';
    case 900:
      return '15';
    case 1_800:
      return '30';
    case 3_600:
      return '60';
    case 14_400:
      return '240';
    case DAY:
      return '1D';
    case WEEK:
      return '1W';
    default:
      return seconds === null ? null : String(seconds);
  }
}

function bucketKey(
  epochSeconds: number,
  seconds: number | null,
  months: number | null,
  weekly: boolean,
): string {
  if (months !== null) return String(Math.floor(monthIndex(epochSeconds) / months));
  if (seconds === null) return String(epochSeconds);
  if (weekly) {
    return String(Math.floor((epochSeconds - MONDAY_EPOCH_OFFSET) / seconds));
  }
  return String(Math.floor(epochSeconds / seconds));
}

function aggregate(
  candles: readonly UsrCandle[],
  seconds: number | null,
  months: number | null,
  weekly: boolean,
  chartSeconds: number | null,
  useChartClock: boolean,
  continuousSession: boolean,
): UsrAnalysisCandle[] {
  const groups: Array<{ key: string; start: number; end: number; candle: UsrCandle }> = [];
  candles.forEach((candle, chartIndex) => {
    const key = bucketKey(candle.time, seconds, months, weekly);
    const last = groups[groups.length - 1];
    if (!last || last.key !== key) {
      groups.push({ key, start: chartIndex, end: chartIndex, candle: { ...candle } });
      return;
    }
    last.end = chartIndex;
    last.candle.high = Math.max(last.candle.high, candle.high);
    last.candle.low = Math.min(last.candle.low, candle.low);
    last.candle.close = candle.close;
    last.candle.volume += candle.volume;
  });

  const chartContext = useChartClock;
  const usable = chartContext ? groups : groups.slice(0, -1);
  const analysis = usable.map((group, index): UsrAnalysisCandle => {
    const next = groups[index + 1];
    const eventChartIndex = chartContext ? group.end : (next?.start ?? group.end);
    return {
      ...group.candle,
      chartStartIndex: group.start,
      chartEndIndex: group.end,
      eventChartIndex,
      eventTime: candles[eventChartIndex].time + (chartSeconds ?? 0),
      closeTime: candles[group.end].time + (chartSeconds ?? 0),
      regularSession: continuousSession || isRegularSession(group.candle.time),
      atr: null,
      volumeMean: null,
      volumeStd: null,
    };
  });
  const atr = atrSeries(analysis);
  analysis.forEach((candle, index) => {
    candle.atr = atr[index];
  });
  return analysis;
}

export function prepareUsrHistory(
  input: readonly UsrCandle[],
  settings: UsrSettings,
  context: UsrComputeContext,
): UsrPreparedHistory {
  const warnings: string[] = [];
  const candlesByTime = new Map<number, UsrCandle>();
  for (const candle of input) {
    if (isFiniteCandle(candle)) candlesByTime.set(candle.time, { ...candle });
  }
  // A later record at the same timestamp is the provider/live-stream
  // correction. Keeping the first record would silently analyze stale OHLCV.
  const sorted = [...candlesByTime.values()].sort((a, b) => a.time - b.time);
  if (sorted.length !== input.length) {
    warnings.push('Invalid or duplicate candles were excluded before analysis.');
  }

  let confirmedCount = sorted.length;
  const last = sorted.at(-1);
  const lastCandleIsOpen =
    context.lastCandleIsOpen ??
    (last !== undefined &&
      (context.chartIntervalSeconds === null ||
        last.time + context.chartIntervalSeconds > context.now));
  if (last && lastCandleIsOpen) {
    confirmedCount -= 1;
  }
  const chartCandles = sorted.slice(0, Math.max(0, confirmedCount));
  const requested = selectedTimeframe(settings, context.chartIntervalSeconds);
  let analysisSeconds = requested?.seconds ?? null;
  let analysisMonths = requested?.months ?? null;
  let usedChartTimeframe = settings.analysisTimeframe === 'chart';
  const requestedComparableSeconds =
    requested?.seconds ?? (requested?.months ? requested.months * (365.25 / 12) * DAY : null);
  const requestedIsChartClock =
    settings.analysisTimeframe === 'chart' ||
    (requested !== null && requested.tag === chartTimeframeTag(context.chartIntervalSeconds));
  const requestedIsLower =
    context.chartIntervalSeconds !== null &&
    requestedComparableSeconds !== null &&
    requestedComparableSeconds < context.chartIntervalSeconds;
  if (
    context.chartIntervalSeconds === null ||
    requested === null ||
    requestedComparableSeconds === null ||
    requestedIsLower ||
    (requestedComparableSeconds === context.chartIntervalSeconds && requestedIsChartClock)
  ) {
    if (requestedIsLower) {
      warnings.push(
        'The selected analysis timeframe is not above the chart timeframe; chart bars are used.',
      );
    }
    analysisSeconds = context.chartIntervalSeconds;
    analysisMonths = null;
    usedChartTimeframe = true;
  }
  const timeframeTag = usedChartTimeframe
    ? 'chart'
    : (requested?.tag ?? analysisSeconds?.toString() ?? 'chart');
  const analysisCandles = aggregate(
    chartCandles,
    analysisSeconds,
    analysisMonths,
    !usedChartTimeframe && timeframeTag.endsWith('W'),
    context.chartIntervalSeconds,
    usedChartTimeframe,
    context.continuousSession === true,
  );
  applyLaggedVolumeBaselines(
    analysisCandles,
    settings.volumeLookback,
    settings.sessionAwareVolume,
    analysisSeconds !== null && analysisSeconds < DAY,
  );
  return {
    presentationCandles: sorted,
    chartCandles,
    analysisCandles,
    analysisSeconds,
    timeframeTag,
    usedChartTimeframe,
    warnings,
  };
}

File: apps/ios/0dteTrader/Features/Chart/UltimateSupportResistance/UsrTimeframe.swift
import Foundation

struct UsrPreparedHistory {
    /// Canonical chart input, including the newest open candle for rendering.
    let presentationCandles: [Candle]
    /// Fully closed chart candles used by every analytical state transition.
    let chartCandles: [Candle]
    let analysisCandles: [UsrAnalysisCandle]
    let analysisSeconds: TimeInterval?
    let timeframeTag: String
    let usedChartTimeframe: Bool
    let warnings: [String]
}

enum UsrTimeframe {
    private static let day: TimeInterval = 86_400
    private static let week: TimeInterval = 604_800
    private static let mondayOffset: TimeInterval = 345_600

    struct Value {
        let seconds: TimeInterval?
        let months: Int?
        let ticks: Int?
        let tag: String
    }

    private static func fixedSeconds(_ seconds: TimeInterval, tag: String? = nil) -> Value {
        Value(seconds: seconds, months: nil, ticks: nil, tag: tag ?? String(Int(seconds)))
    }

    private static func fixedMonths(_ months: Int) -> Value {
        Value(seconds: nil, months: months, ticks: nil, tag: "\(months)M")
    }

    private static func fixedTicks(_ ticks: Int) -> Value {
        Value(seconds: nil, months: nil, ticks: ticks, tag: "\(ticks)T")
    }

    private static func auto(_ chart: TimeInterval) -> Value {
        if chart <= 5 { return fixedSeconds(15, tag: "15S") }
        if chart <= 15 { return fixedSeconds(60, tag: "1") }
        if chart <= 30 { return fixedSeconds(120, tag: "2") }
        if chart <= 60 { return fixedSeconds(300, tag: "5") }
        if chart <= 180 { return fixedSeconds(900, tag: "15") }
        if chart <= 300 { return fixedSeconds(1_800, tag: "30") }
        if chart <= 900 { return fixedSeconds(3_600, tag: "60") }
        if chart <= 1_800 { return fixedSeconds(7_200, tag: "120") }
        if chart <= 3_600 { return fixedSeconds(14_400, tag: "240") }
        if chart <= 7_200 { return fixedSeconds(28_800, tag: "480") }
        if chart <= 14_400 { return fixedSeconds(day, tag: "1D") }
        if chart <= 43_200 { return fixedSeconds(2 * day, tag: "2D") }
        if chart <= day { return fixedSeconds(week, tag: "1W") }
        if chart <= 3 * day { return fixedSeconds(2 * week, tag: "2W") }
        if chart <= week { return fixedMonths(1) }
        if chart <= 2 * week { return fixedMonths(2) }
        if chart <= 31 * day { return fixedMonths(3) }
        return fixedMonths(12)
    }

    static func parse(_ input: String) -> Value? {
        let value = input.trimmingCharacters(in: .whitespacesAndNewlines).uppercased()
        guard let match = value.range(of: #"^([1-9][0-9]*)?(T|S|D|W|M)?$"#,
                                      options: .regularExpression) else { return nil }
        let canonical = String(value[match])
        guard !canonical.isEmpty else { return nil }
        let suffix = canonical.last.map(String.init) ?? ""
        let hasSuffix = ["T", "S", "D", "W", "M"].contains(suffix)
        let digits = hasSuffix ? String(canonical.dropLast()) : canonical
        guard let amount = Int(digits.isEmpty ? "1" : digits) else { return nil }
        switch hasSuffix ? suffix : "" {
        case "T":
            return [1, 10, 100, 1_000].contains(amount) ? fixedTicks(amount) : nil
        case "S":
            return [1, 5, 10, 15, 30, 45].contains(amount)
                ? fixedSeconds(Double(amount), tag: "\(amount)S") : nil
        case "D":
            return amount <= 365 ? fixedSeconds(Double(amount) * day, tag: "\(amount)D") : nil
        case "W":
            return amount <= 52 ? fixedSeconds(Double(amount) * week, tag: "\(amount)W") : nil
        case "M":
            return amount <= 12 ? fixedMonths(amount) : nil
        default:
            return amount <= 1_440 ? fixedSeconds(Double(amount) * 60, tag: canonical) : nil
        }
    }

    private static func selected(
        _ settings: UsrSettings,
        chart: TimeInterval?
    ) -> Value? {
        switch settings.analysisTimeframe {
        case "chart": return chart.map { fixedSeconds($0, tag: chartTag($0)) }
        case "auto": return chart.map(auto)
        case "4h": return fixedSeconds(14_400, tag: "240")
        case "1d": return fixedSeconds(day, tag: "1D")
        case "3d": return fixedSeconds(3 * day, tag: "3D")
        case "1w": return fixedSeconds(week, tag: "1W")
        case "2w": return fixedSeconds(2 * week, tag: "2W")
        case "1m": return fixedMonths(1)
        case "custom": return parse(settings.customTimeframe)
        default: return nil
        }
    }

    private static func regularSession(_ date: Date) -> Bool {
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = TimeZone(identifier: "America/New_York") ?? .current
        let weekday = calendar.component(.weekday, from: date)
        guard weekday != 1 && weekday != 7 else { return false }
        let minute = calendar.component(.hour, from: date) * 60 + calendar.component(.minute, from: date)
        return minute >= 570 && minute < 960
    }

    private static func chartTag(_ seconds: TimeInterval?) -> String? {
        guard let seconds else { return nil }
        if seconds == 60 { return "1" }
        if seconds == 300 { return "5" }
        if seconds == 900 { return "15" }
        if seconds == 1_800 { return "30" }
        if seconds == 3_600 { return "60" }
        if seconds == 14_400 { return "240" }
        if seconds == day { return "1D" }
        if seconds == week { return "1W" }
        return String(Int(seconds))
    }

    private static func key(
        _ date: Date,
        seconds: TimeInterval?,
        months: Int?,
        weekly: Bool
    ) -> String {
        if let months {
            var calendar = Calendar(identifier: .gregorian)
            calendar.timeZone = TimeZone(secondsFromGMT: 0) ?? .current
            let monthIndex = calendar.component(.year, from: date) * 12
                + calendar.component(.month, from: date) - 1
            return String(monthIndex / months)
        }
        let epoch = date.timeIntervalSince1970
        guard let seconds else { return String(epoch) }
        if weekly {
            return String(Int(floor((epoch - mondayOffset) / seconds)))
        }
        return String(Int(floor(epoch / seconds)))
    }

    private struct Group {
        let key: String
        let start: Int
        var end: Int
        var candle: Candle
    }

    static func prepare(
        candles input: [Candle],
        settings: UsrSettings,
        chartSeconds: TimeInterval?,
        continuousSession: Bool = false,
        now: Date,
        lastCandleIsOpen: Bool? = nil
    ) -> UsrPreparedHistory {
        var warnings: [String] = []
        var candlesByTime: [Date: Candle] = [:]
        for candle in input where candle.time.timeIntervalSince1970.isFinite
            && candle.open.isFinite && candle.high.isFinite
            && candle.low.isFinite && candle.close.isFinite && candle.volume.isFinite
            && candle.volume >= 0 && candle.high >= candle.low
            && candle.open >= candle.low && candle.open <= candle.high
            && candle.close >= candle.low && candle.close <= candle.high {
            // Prefer the last valid provider/live-stream correction for a
            // timestamp instead of retaining stale OHLCV from the first copy.
            candlesByTime[candle.time] = candle
        }
        let sorted = candlesByTime.values.sorted { $0.time < $1.time }
        if sorted.count != input.count { warnings.append("Invalid or duplicate candles were excluded before analysis.") }
        var confirmed = sorted
        let newestIsOpen = lastCandleIsOpen ?? confirmed.last.map {
            chartSeconds == nil || $0.time.addingTimeInterval(chartSeconds ?? 0) > now
        } ?? false
        if newestIsOpen, !confirmed.isEmpty {
            confirmed.removeLast()
        }
        let requested = selected(settings, chart: chartSeconds)
        var seconds = requested?.seconds
        var months = requested?.months
        var chartContext = settings.analysisTimeframe == "chart"
        let comparable: TimeInterval?
        if let requestedSeconds = requested?.seconds {
            comparable = requestedSeconds
        } else if let requestedMonths = requested?.months {
            comparable = Double(requestedMonths) * (365.25 / 12) * day
        } else {
            comparable = nil
        }
        let requestedIsChartClock = settings.analysisTimeframe == "chart"
            || requested?.tag == chartTag(chartSeconds)
        let requestedIsLower = chartSeconds != nil && comparable != nil
            && (comparable ?? 0) < (chartSeconds ?? 0)
        if chartSeconds == nil || requested == nil || comparable == nil || requestedIsLower
            || (comparable == chartSeconds && requestedIsChartClock) {
            if requestedIsLower {
                warnings.append("The selected analysis timeframe is not above the chart timeframe; chart bars are used.")
            }
            seconds = chartSeconds
            months = nil
            chartContext = true
        }
        let timeframeTag = chartContext ? "chart" : requested?.tag ?? "chart"
        let weeklyClock = !chartContext && timeframeTag.hasSuffix("W")
        var groups: [Group] = []
        for (index, candle) in confirmed.enumerated() {
            let bucket = key(candle.time, seconds: seconds, months: months, weekly: weeklyClock)
            if groups.last?.key != bucket {
                groups.append(Group(key: bucket, start: index, end: index, candle: candle))
            } else {
                let last = groups.count - 1
                groups[last].end = index
                groups[last].candle.high = max(groups[last].candle.high, candle.high)
                groups[last].candle.low = min(groups[last].candle.low, candle.low)
                groups[last].candle.close = candle.close
                groups[last].candle.volume += candle.volume
            }
        }
        let isChartClock = chartContext
        let usable = isChartClock ? groups : Array(groups.dropLast())
        var analysis = usable.enumerated().map { index, group in
            let eventChartIndex = isChartClock ? group.end : groups[index + 1].start
            return UsrAnalysisCandle(
                time: group.candle.time,
                open: group.candle.open,
                high: group.candle.high,
                low: group.candle.low,
                close: group.candle.close,
                volume: group.candle.volume,
                chartStart: group.start,
                chartEnd: group.end,
                eventChartIndex: eventChartIndex,
                eventTime: confirmed[eventChartIndex].time.addingTimeInterval(chartSeconds ?? 0),
                closeTime: confirmed[group.end].time.addingTimeInterval(chartSeconds ?? 0),
                regularSession: continuousSession || regularSession(group.candle.time),
                atr: nil,
                volumeMean: nil,
                volumeStd: nil
            )
        }
        let aggregateCandles = analysis.map {
            Candle(time: $0.time, open: $0.open, high: $0.high, low: $0.low, close: $0.close, volume: $0.volume)
        }
        let atr = UsrMath.atr(aggregateCandles)
        var all: [Double] = []
        var regular: [Double] = []
        var extended: [Double] = []
        for index in analysis.indices {
            analysis[index].atr = atr[index]
            let chosen = settings.sessionAwareVolume && (seconds ?? day) < day
                ? (analysis[index].regularSession ? regular : extended)
                : all
            let fallback: ArraySlice<Double>? = all.count >= settings.volumeLookback
                ? all.suffix(settings.volumeLookback) : nil
            let sample: ArraySlice<Double>? = chosen.count >= settings.volumeLookback
                ? chosen.suffix(settings.volumeLookback) : fallback
            if let sample {
                let mean = UsrMath.mean(sample)
                analysis[index].volumeMean = mean
                analysis[index].volumeStd = UsrMath.deviation(sample, mean: mean)
            }
            all.append(analysis[index].volume)
            if analysis[index].regularSession { regular.append(analysis[index].volume) }
            else { extended.append(analysis[index].volume) }
        }
        return UsrPreparedHistory(
            presentationCandles: sorted,
            chartCandles: confirmed,
            analysisCandles: analysis,
            analysisSeconds: seconds,
            timeframeTag: timeframeTag,
            usedChartTimeframe: chartContext,
            warnings: warnings
        )
    }
}
